architecture/2024-10-deploymenttemplate-controller.md

```diff
@@ -0,0 +1,421 @@
+
+**Goal: The behavior of DeploymentTemplate is consistent with user expectations for a Kubernetes-enabled tool**
+- We will follow user expectations for Kubernetes controllers and resources when designing the DeploymentTemplate controller and resource. This includes building in support for retries, status updates, and error handling.
+
+### Non-goals
+
+**Non-goal (out of scope): Full support for GitOps**
+
+- We will not yet be implementing automatic generation of DeploymentTemplate resources from Bicep manifests or querying Git repositories. This design will enable this work, and it will be covered in a future design document.
+
+### User scenarios
+
+#### Jon can deploy cloud (Azure or AWS) resources defined in Bicep manifests using Kubernetes tools
+
+Jon is an infrastructure operator for an enterprise company. His team manages a production application that is deployed on Kubernetes and uses cloud resources. Jon wants to declare all dependencies for his application in Kubernetes manifests, including cloud resources while leveraging Radius for its useful features. He installs Radius on his cluster and uses the rad CLI to generate a custom resource from his Bicep manifest. Jon applies the custom resource to his cluster, and Radius deploys the cloud resources defined in the Bicep manifest. If he wants to update or delete the cloud resources, he can do so by re-applying the updated custom resource or deleting the custom resource.
+
+#### Jon can deploy Radius resources defined in Bicep manifests using Kubernetes tools
+
+Now that he can see that Radius can deploy cloud resources defined in Bicep manifests, Jon wants to take advantage of Radius tooling, such as the App Graph, and fully "Rad-ify" his application. He writes a Bicep manifest that defines a Radius container that connects to the cloud resources, and uses the rad CLI to generate a custom resource from the Bicep manifest. Jon applies the custom resource to his cluster, and Radius deploys the Radius resources defined in the Bicep manifest. Now, Jon can take advantage of Radius tooling, such as the Radius Dashboard and App Graph, to manage his application.
+
+## User Experience
+
+### Radius Container
+
+**Sample Input:**
+
+#### `app.bicep`
+```bicep
+extension radius
+
+param port int
+param tag string
+
+resource demoenv 'Applications.Core/environments@2023-10-01-preview' existing = {
+  name: 'demoenv'
+}
+
+resource demoapp 'Applications.Core/applications@2023-10-01-preview' = {
+  name: 'demoapp'
+  properties: {
+    environment: demoenv.id
+  }
+}
+
+resource democtnr 'Applications.Core/containers@2023-10-01-preview' = {
+  name: 'democtnr'
+  properties: {
+    application: demoapp.id
+    container: {
+      image: 'ghcr.io/radius-project/samples/demo:${tag}'
+      ports: {
+        web: {
+          containerPort: port
+        }
+      }
+    }
+  }
+}
+```
```

Comment:
How will this work when bicep file app.bicep refers to another bicep module? Is that within the scope of this feature at this time?

Reply:
Yes, this design should cover this scenario because we rely on the Bicep compiler to handle these cases.